sdk/core/System.ClientModel/src/Pipeline/ClientRetryPolicy.cs

```diff
@@ -0,0 +1,238 @@
+            }
+
+            // ShouldRetry returned false this iteration and the last request
+            // we sent didn't cause an exception to be thrown.
+            // So, we're done.  Exit the while loop.
+            break;
+        }
+    }
+
+    protected virtual void OnSendingRequest(PipelineMessage message) { }
+
+    protected virtual ValueTask OnSendingRequestAsync(PipelineMessage message) => default;
+
+    protected virtual void OnRequestSent(PipelineMessage message) { }
+
+    protected virtual ValueTask OnRequestSentAsync(PipelineMessage message) => default;
+
+    protected virtual void OnTryComplete(PipelineMessage message) { }
+
+    internal bool ShouldRetry(PipelineMessage message, Exception? exception)
+    {
+        // If there was no exception and we got a success response, don't retry.
+        if (exception is null && message.Response is not null && !message.Response.IsError)
+        {
+            return false;
+        }
+
+        return ShouldRetryCore(message, exception);
+    }
+
+    internal async ValueTask<bool> ShouldRetryAsync(PipelineMessage message, Exception? exception)
+    {
+        // If there was no exception and we got a success response, don't retry.
+        if (exception is null && message.Response is not null && !message.Response.IsError)
+        {
+            return false;
+        }
+
+        return await ShouldRetryCoreAsync(message, exception).ConfigureAwait(false);
+    }
+
+    protected virtual bool ShouldRetryCore(PipelineMessage message, Exception? exception)
+    {
+        if (message.RetryCount >= _maxRetries)
+        {
+            // We've exceeded the maximum number of retries, so don't retry.
+            return false;
+        }
+
+        return exception is null ?
+            IsRetriable(message) :
+            IsRetriable(message, exception);
+    }
+
+    protected virtual ValueTask<bool> ShouldRetryCoreAsync(PipelineMessage message, Exception? exception)
+        => new(ShouldRetryCore(message, exception));
+
+    internal TimeSpan GetNextDelay(PipelineMessage message, int tryCount)
+    => GetNextDelayCore(message, tryCount);
+
+    protected virtual TimeSpan GetNextDelayCore(PipelineMessage message, int tryCount)
+    {
+        // Default implementation is exponential backoff
+        return TimeSpan.FromMilliseconds((1 << (tryCount - 1)) * _initialDelay.TotalMilliseconds);
+    }
+
+    protected virtual async Task WaitAsync(TimeSpan time, CancellationToken cancellationToken)
+    {
+        await Task.Delay(time, cancellationToken).ConfigureAwait(false);
+    }
+
+    protected virtual void Wait(TimeSpan time, CancellationToken cancellationToken)
+    {
+        cancellationToken.WaitHandle.WaitOne(time);
+    }
+
+    #region Retry Classifier
+
+    // Overriding response-retriable classification will be added in a later ClientModel release.
```

Comment:
But aren't we already adding the MessageClassifier in _this_ release?

Reply:
Nevermind, I see that MessageClassifier doesn't yet have the Retriable API.

Reply:
Yep, that's what the comment was addressing.  I'm interested in seeing if we can separate retry classifiers from error classifiers when we add this to the design, but we can address it later I think.